File: Python/Problem039.py
"""
Project Euler, problem 39

If p is the perimeter of a right angle triangle with integral length sides,
{a,b,c}, there are exactly three solutions for p = 120.

{20,48,52}, {24,45,51}, {30,40,50}

For which value of p <= 1000, is the number of solutions maximized?

answer = 840;  //correct
"""

import math

perimSolutions = []
answer = 0

for x in range(1000):
    perimSolutions.append(0)

for p in range(1, 1001):

    for a in range(1, int((p/2) + 1)):
        for b in range(a + 1, int((p /2) + 1)):
            c = math.sqrt(a*a + b*b)
            if a + b + c == p:
                print(a, ", ", b, ", ", c )
                perimSolutions[p - 1] += 1

for i in range(1000):
    if perimSolutions[i] > perimSolutions[answer]:
        answer = i

print(answer + 1)

File: Java/src/Problem039.java
/* Project Euler, problem 39
 * 
 * If p is the perimeter of a right angle triangle with integral length sides,
 * {a,b,c}, there are exactly three solutions for p = 120.
 * 
 * {20,48,52}, {24,45,51}, {30,40,50}
 * 
 * For which value of p <= 1000, is the number of solutions maximized?
 * 
 * answer = 840;  //correct
 */



public class Problem039 {
	public static void main(String[] args){
		
		int[] perimSolutions = new int[1000];
		int answer = 0;
		
		for (int p = 1; p <= 1000; p++) {
			
			for (int a = 1; a <= p / 2; a++) {
				for (int b = a + 1; b <= p / 2; b++) {
					double c = Math.sqrt(a*a + b*b);
					if(a + b + c == p){
						perimSolutions[p - 1]++;
					}
				}
			}
		}
		for (int i = 0; i < 1000; i++) {
			if(perimSolutions[i] > perimSolutions[answer]){
				answer = i;
			}
		}
		System.out.println("perimeter: " + (answer + 1) + " has " + perimSolutions[answer] + " solutions.");
		
		
	}
}
